### PR TITLE
consensus/istanbul: Istanbul BFT (EIP650)

### PR DESCRIPTION
**Pull request by yutelin**
This PR is the implementation of Istanbul BFT protocol in [ethereum/EIPs#650](https://github.com/ethereum/EIPs/issues/650)

### CODE REVIEW

**Comment by fjl** [comment]
Warning: it will take a while to get this merged ;)

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 2620c98d1058bcb7d6c13e848737fddf45823abe
  * Commits must be prefixed with the package(s) they modify

 * Commit: 7e94f76b639180609ce99c46b2a7e1921fb50991
  * Commits must be prefixed with the package(s) they modify

 * Commit: 3f184312489f6ca1c3c1565fb677e3b1ab8e1d41
  * Commits must be prefixed with the package(s) they modify

 * Commit: 687d5ca084d37a97b84feec7b823b71d8232f938
  * Commits must be prefixed with the package(s) they modify

 * Commit: edcbde333638bf5b22149d582ea33474da19decc
  * Commits must be prefixed with the package(s) they modify

 * Commit: bd88b3e41622a10fae05a7f3338a559aff820a16
  * Commits must be prefixed with the package(s) they modify

 * Commit: 165e40bbd87b318708bbcd75a27f6361f292c5a6
  * Commits must be prefixed with the package(s) they modify

 * Commit: 214c4e1254036bc198b73f549142b3cc5189510e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 66374980ded687e740873479a3f08103cf2a4000
  * Commits must be prefixed with the package(s) they modify

 * Commit: ef7b0fc164608c2096485bf38434abe7dc6a0660
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1f7da3472194296933edbce7264c9daf0e204d97
  * Commits must be prefixed with the package(s) they modify

 * Commit: f81f870eaceae38b5c0f0af6497df686561e8cd5
  * Commits must be prefixed with the package(s) they modify

 * Commit: 13c4671286a0c80157d23fdf3e5ca54d5e7bd058
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1550260758645de6a262fd826b97ae22817d46ff
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by stevenroose** [comment]
`--istanbul.blockpausetime value  Pause time when zero tx in previous block, values should be larger than istanbul.blockperiod (default: 2)`
I guess it's a good idea to fail when that is not the case :)

File: consensus/istanbul/backend/backend_test.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	block := makeBlock(chain, engine, chain.Genesis())
+	chain.InsertChain(types.Blocks{block})
+	expected := engine.GetProposer(1)
+	actual := engine.Address()
+	if actual != expected {
+		t.Errorf("proposer mismatch: have %v, want %v", actual.Hex(), expected.Hex())
+	}
+}
+
+/**
```
Comment (stevenroose) [inline consensus/istanbul/backend/backend_test.go]: Forgotten comment?

**Comment by kesar** [comment]
could conflicts be solved? I would love to use this algo!

**Comment by stevenroose** [comment]
-- repost of [comment on EIP 650](https://github.com/ethereum/EIPs/issues/650#issuecomment-360085474) --

The current implementation (as found in Quorum) breaks the concept of the "pending" block, used in several calls, but most notably in `eth_getTransactionCount` ([`PendingNonceAt`](https://godoc.org/github.com/obscuren/go-ethereum/ethclient#Client.PendingNonceAt) in `ethclient`):

In Ethereum, the pending block means the latest confirmed block + all pending transactions the node is aware of. This means that directly after a transaction is sent to the node (through RPC), the transaction count (aka nonce) in the "pending" block is increased. A lot of tools, like abigen in this repo or any other tool where tx signing occurs at the application level instead of in geth, rely on this for making multiple transactions at once. After the first one, the result of `eth_getTransactionCount` will increase so that a valid second tx can be crafted.

With the current implementation of Istanbul, the definition of the "pending block" seem to be different. When submitting a transaction, the result for `eth_getTransactionCount` for the sender in the "pending" block does not change. When a new block is confirmed (not containing this tx), it does change however (while the value for "latest" doesn't). Then, on the next block confirmation, the "latest" also changes because the tx is in the confirmed block.

So this seems to mean that the "pending block" definition changed from "latest block + pending txs" to "the block that is currently being voted on". I consider this a bug; if this is done on purpose, it breaks with a lot of existing applications (all users of abigen, f.e.) and should be reconsidered.

I originally reported about this issue [in the Quorum repo](https://github.com/jpmorganchase/quorum/issues/247), but there doesn't seem to be a good place to report bugs in Istanbul other than here.

File: consensus/istanbul/core/handler.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+func (c *core) Stop() error {
+	c.stopTimer()
+	c.unsubscribeEvents()
+
+	// Make sure the handler goroutine exits
+	c.handlerWg.Wait()
+	return nil
+}
+
+// ----------------------------------------------------------------------------
```
Comment (AntoineRondelet) [inline consensus/istanbul/core/handler.go]: I would remove this comment

File: consensus/istanbul/core/message_set.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+			Round:    new(big.Int),
+			Sequence: new(big.Int),
+		},
+		messagesMu: new(sync.Mutex),
+		messages:   make(map[common.Address]*message),
+		valSet:     valSet,
+	}
+}
+
+// ----------------------------------------------------------------------------
```
Comment (AntoineRondelet) [inline consensus/istanbul/core/message_set.go]: Useless comment ?

File: consensus/istanbul/core/message_set.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	return len(ms.messages)
+}
+
+func (ms *messageSet) Get(addr common.Address) *message {
+	ms.messagesMu.Lock()
+	defer ms.messagesMu.Unlock()
+	return ms.messages[addr]
+}
+
+// ----------------------------------------------------------------------------
```
Comment (AntoineRondelet) [inline consensus/istanbul/core/message_set.go]: Useless comment ?

File: consensus/istanbul/core/roundstate.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+		Commits:        newMessageSet(validatorSet),
+		lockedHash:     lockedHash,
+		mu:             new(sync.RWMutex),
+		pendingRequest: pendingRequest,
+		hasBadProposal: hasBadProposal,
+	}
+}
+
+// roundState stores the consensus state
+type roundState struct {
```
Comment (AntoineRondelet) [inline consensus/istanbul/core/roundstate.go]: I would move this struct just before the `newRoundState` function

**Comment by kaidiren** [comment]
cool, please go ahead

**Comment by adamschmideg** [comment]
We don't have the capacity to merge and maintain it. We had an informal chat with Quorum, and they might write the integration.

**Comment by d4nd** [comment]
Is there any update regarding the integration?

**Comment by jpmsam** [comment]
It is integrated in Quorum which is a fork of Geth.

**Comment by d4nd** [comment]
> It is integrated in Quorum which is a fork of Geth.

Yes, I'm aware of it.
But I mean when it's going to be merged into Ethereum geth official repository.

File: consensus/istanbul/backend/engine.go [hunk 6]
```diff
@@ -0,0 +1,717 @@
...
+}
+
+// writeCommittedSeals writes the extra-data field of a block header with given committed seals.
+func writeCommittedSeals(h *types.Header, committedSeals [][]byte) error {
+	if len(committedSeals) == 0 {
+		return errInvalidCommittedSeals
+	}
+
+	for _, seal := range committedSeals {
+		if len(seal) != types.IstanbulExtraSeal {
```
Comment (nambrot) [inline consensus/istanbul/backend/engine.go]: Could someone tell me what this checks? I was trying to figure out where we check during the commit/finalization that the commit seals are valid?

**Comment by 0xrebased** [comment]
wen merge >:(

**Comment by karalabe** [comment]
never

**Comment by zhiqiangxu** [comment]
> never

@karalabe  Is there an official reason for this?

**Comment by sarthaktNt** [comment]
Polygon Edge migration from POA to POS
 
I have created a 4 nodes POA network using polygon-edge.
As mentioned ,I have tried out the approach for migration.

I have deployed the staking smart contract in block no 59357 .All the validators staking has staked before block no 59750.I have verified the list in staking contract.
I have restarted all the nodes after migration to rebuild the blocks from new genesis.json.
polygon-edge ibft switch --chain ./genesis.json --type PoS --deployment 59357 --from 59750
But the blocks are not getting generated from 59750 due to the following error.

**failed to get validators when calculation quorum**


Please guys help me to resolve my issue

**Comment by karalabe** [comment]
This is a repo for Ethereum, not Polygon.

**Pull request closed without merging**